### PR TITLE
Re-adding ability to specify consistency levels at query level for CQ…

### PR DESCRIPTION
**Pull request by alexflav23**
…L protocol V3.

- Adding per-protocol version logic to specify consistency levels.
- Version 2 and below consistency is specified in an `USING` clause by phantom.
- Consistency levels for V3 of the protocol are simply being forwarded to the Datastax Java Driver via `SimpleStatement` to allow the driver to set the level internally outside of phantom.
- `consistencyLevel` argument is now propagated everywhere.
- Added tests to make sure the consistency level is propagated in settings when the latest protocol is used.

### CODE REVIEW

File: phantom-dsl/src/main/scala/com/websudos/phantom/builder/query/prepared/PreparedUpdateQuery.scala [hunk 1]
```diff
@@ -163,16 +173,32 @@ sealed class AssignmentsQuery[
...
-  def consistencyLevel_=(level: ConsistencyLevel)(implicit ev: Status =:= Unspecified): AssignmentsQuery[Table, Record, Limit, Order, Specified, Chain] = {
-    new AssignmentsQuery(
-      table,
-      init,
-      usingPart append QueryBuilder.consistencyLevel(level.toString),
-      wherePart,
-      setPart,
-      casPart
-    )
+  def consistencyLevel_=(level: ConsistencyLevel)(implicit ev: Status =:= Unspecified, session: Session): AssignmentsQuery[Table, Record, Limit, Order, Specified, Chain] = {
```
Comment (bjankie1) [inline phantom-dsl/src/main/scala/com/websudos/phantom/builder/query/prepared/PreparedUpdateQuery.scala]: The only information required from session is the protocol version, therefore it'd be cleaner just to pass this as argument. Implicit conversion should help with avoiding boiler plate required to extract it from session

File: phantom-dsl/src/main/scala/com/websudos/phantom/builder/query/prepared/PreparedUpdateQuery.scala [hunk 2]
```diff
@@ -163,16 +173,32 @@ sealed class AssignmentsQuery[
...
+    val protocol = session.getCluster.getConfiguration.getProtocolOptions.getProtocolVersionEnum
+
+    if (protocol.compareTo(ProtocolVersion.V2) == 1) {
```
Comment (bjankie1) [inline phantom-dsl/src/main/scala/com/websudos/phantom/builder/query/prepared/PreparedUpdateQuery.scala]: Enums can be compared using equality `==` operator

File: phantom-dsl/src/main/scala/com/websudos/phantom/builder/query/prepared/PreparedUpdateQuery.scala [hunk 3]
```diff
@@ -163,16 +173,32 @@ sealed class AssignmentsQuery[
...
+        wherePart,
+        setPart,
+        casPart,
+        level
+      )
+    } else {
+      new AssignmentsQuery(
+        table,
+        init,
+        usingPart append QueryBuilder.consistencyLevel(level.toString),
```
Comment (bjankie1) [inline phantom-dsl/src/main/scala/com/websudos/phantom/builder/query/prepared/PreparedUpdateQuery.scala]: The level can be null at this point. Is `consistencyLevel` setter mandatory? If not why not to make the level field an `Option[ConsistencyLevel]` and call the append method only when the level is not `None`?

**Comment by nosheenzaza** [comment]
Hello Flavian,

I am not sure why you attach the consistency level to a statement only for protocol V3 or above and pass 'null' otherwise to still use USING clause for older protocol versions. I was checking [com.datastax.driver.core.SessionManager](https://github.com/datastax/java-driver/blob/2.1/driver-core/src/main/java/com/datastax/driver/core/SessionManager.java) as well as [com.datastax.driver.core.Requests](https://github.com/datastax/java-driver/blob/2.1/driver-core/src/main/java/com/datastax/driver/core/Requests.java) and it seems to me that the Java driver will internally parse statements and their attached consistency level (regular and serial), then generate a suitable request depending on the protocol internally, this is true not only for consistency but for many newly added features. So I believe there is no need to generate different CQL based on the protocol version, because this will be handled by the driver anyway. 

I think I would always set the consistency level for SimpleStatement regardless of the protocol version,  never generate CQL with consistency set in the USING clause and rely on the Java driver to generate the correct request. Do you think this should work or am I missing something here?

**Comment by alexflav23** [comment]
Hi @nosheenzaza,

Thank you for taking the time to review this in depth. We have since upgraded this to a more idiomatic `Option` based DSL to eliminate having to deal with `null` in the code.

The reason for this is because we don't actually use the Java Driver for any of the serialization, or at the very least we didn't until this requirement came about. At this point I am uncertain whether something like the following will be automatically coerced to the right approach based on protocol version:

```scala

// In theory the Phantom builder would output a string query in the following manner.
val query = "UPDATE keyspace.TABLE SET a = b USING CONSISTENCY LEVEL QUORUM" 

// then in ExecutableQuery.scala we transform this to a statement
val executable = session.newSimpleStatement(query)
```

However it doesn't seem that will do any parsing. If we explicitly call `setConsistencyLevel`, things work as expected:

```scala
val executable = session.newSimpleStatement(query).setConsistencyLevel(level)
```

That's why the new parameter is everywhere. Hope that makes sense but it may indeed be unnecessary, at the time perhaps it wasn't 100% crystal clear where the driver will "take over".

**Pull request closed without merging**